feat: various

zeruth opened the pull request:
adds OpObj,OpLoc,OpNpc,OpPlayer decoders
handle IfButton decoding properly
adds packet g2_alt1/2 
adds packet g4_alt1/2/3

(Read Login Packet)https://github.com/LostCityRS/Engine/pull/5/commits/2a22c3eb051ef4d4b3083a91d0ae1f2878beedc2
(Fix Music Tab)https://github.com/LostCityRS/Engine/pull/5/commits/86a60a7306dcd6c3809ce06318895bc1051480ab

zeruth committed:
feat: update decoders

adds OpPlayer1-8
handle IfButton properly

zeruth committed:
feat: opnpc decoders

adds OpNpc1-5 decoders

zeruth committed:
feat: oploc decoders

adds OpLoc1-5 decoders

zeruth committed:
feat: opobj decoders

adds opobj1-5 decoders

adds packet g4_alt1/2/3

zeruth committed:
fix: packet g2_alt1/2

alt1 was incorrect, alt2 verified

zeruth committed:
fix: cleanup packet additions

zeruth committed:
feat: add examine decoders

adds loc,npc,obj examine decoders
fixes g2_alt2

--easter egg--
adds lumby door knock knock op1
adds lumby door whos there examine

zeruth committed:
chore: rename

Pazaz commented on src/network/os1/client/handler/game/LocExamineHandler.ts:
```diff
@@ -0,0 +1,17 @@
+
+        //todo: remove easter egg
+        if (message.id === 1530) {
+            player.write(new MessageGame('Who\'s there?'));
```
can use double-quotes to single-quote inside :)
"Who's there?"

Pazaz commented on src/network/os1/client/prot/game/GameClientRepository.ts:
```diff
@@ -30,5 +82,31 @@ export default class GameClientRepository extends ClientRepository {
         this.bind(new ClientCheatDecoder(), new ClientCheatHandler())
         this.bind(new IfButtonDecoder(), new IfButtonHandler())
         this.bind(new ClanDecoder(), new ClanHandler())
+        this.bind(new OpPlayer1Decoder(), new OpPlayer1Handler())
```
I would make all of these into single decoder/handlers per *type* (OpPlayer, OpNpc, OpLoc, OpPlayerU, OpNpcU, OpPlayerT, ...), you can pass the opcode into the decoder and track "option" / "op" in the message.

I can give an example on how to implement if you need. Your message model becomes:
```ts
class OpPlayer {
    op: number;
    idx: number;
}
```
And you use that op in the handler.

zeruth replied:
I think I got it the way you wanted. cleaned up the PR quite a bit so thank you

zeruth committed:
feat: cleanup op handlers

Pazaz commented on src/network/os1/client/prot/game/GameClientRepository.ts:
```diff
@@ -30,5 +63,31 @@ export default class GameClientRepository extends ClientRepository {
         this.bind(new ClientCheatDecoder(), new ClientCheatHandler())
         this.bind(new IfButtonDecoder(), new IfButtonHandler())
         this.bind(new ClanDecoder(), new ClanHandler())
+        this.bind(new OpPlayer1Decoder(), new OpPlayerHandler())
```
Much better. You could do the same for the decoders to reduce some boilerplate too! Pass op and opcode into "OpPlayerDecoder":

```ts
this.bind(new OpPlayerDecoder(1, 245), new OpPlayerHandler());
this.bind(new OpPlayerDecoder(2, 146), new OpPlayerHandler());
this.bind(new OpPlayerDecoder(3, 102), new OpPlayerHandler());
this.bind(new OpPlayerDecoder(4, 78), new OpPlayerHandler());
```

```ts
export default class OpPlayerDecoder extends MessageDecoder {
    size = 2;
    limit = GameClientLimit.CLIENT;

    constructor(op: number, opcode: number) {
        this.op = op;
        this.opcode = opcode;
    }

    read(buf: Packet): ClientMessage {
        let playerId = -1;
        if (this.op === 1) {
            playerId = buf.g2();
        } else if (this.op === 2) {
            playerId = buf.g2();
        } else if (this.op === 3) {
            playerId = buf.g2_alt1();
        }

        return new OpPlayer(this.op, playerId);
    }
}
```

Pazaz commented on src/network/os1/client/prot/game/GameClientRepository.ts:
```diff
@@ -30,5 +63,31 @@ export default class GameClientRepository extends ClientRepository {
+        this.bind(new OpObj3Decoder(), new OpObjHandler())
+        this.bind(new OpObj4Decoder(), new OpObjHandler())
+        this.bind(new OpObj5Decoder(), new OpObjHandler())
+        this.bind(new LocExamineDecoder(), new LocExamineHandler())
```
The examine packets are the 6th option with a special engine-case, really called OPLOC6 OPNPC6 OPOBJ6

zeruth committed:
chore: cleanup decoders

zeruth committed:
chore: cleanup imports

zeruth committed:
chore: refactor examine decoders

zeruth commented:
Those issues are resolved as well. sorry for not seeing the decoder side of cleanup. I get what you meant now.
Overall the PR is looking much cleaner and does the same job, thanks again!

zeruth committed:
feat: read login packet

this is done the way the client is currently configured to connect. naturally this will need to be updated when the rsa/isaac standards improve

zeruth committed:
chore: cleanup

remove pointless easter egg lol
remove console log on if button

zeruth committed:
feat: music tab

zeruth commented:
Did some more cleanup and added a bonus of doing the login packet and fixing the music tab :) I won't be adding to this PR minus any more feedback
